src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Array.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Buffers.Binary;
+using System.Text;
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal partial class CborWriter
+    {
+        public void BeginWriteArray(int definiteLength)
```

Comment:
Should there be a story for the writer supporting the counting when the caller doesn't?  E.g. if the caller is draining an IEnumerable the only safe way to count and re-enumerate is to drain it to a list.

For arrays of length < 27 (assuming I remember the constant correctly) the length can be updated without shifting data on EndArray.

Reply:
(While indefinite would support that, I mean "without using an indefinite encoding")

Reply:
What's the motivation for supporting this? Eliminating the trailing byte? Optimizing reads?

> For arrays of length < 27 (assuming I remember the constant correctly) the length can be updated without shifting data on EndArray.

Are we prepared to shift bytes if the array is large enough? If not, knowing that an IEnumerable has less that 24 elements requires partially enumerating it, before any elements get written to the buffer. So I think adding this might encourage weird usage patterns where IEnumerators and intermediate collections are being handled explicitly. It will also complicate Stream support.

I'd be inclined to say we should not support this, and encourage users to use the indefinite-length method if it's not known from the start.

Reply:
> What's the motivation for supporting this? Eliminating the trailing byte?

Supporting the restricted canonicalization modes. (Since our sponsoring use case uses a canonical mode that prohibits indefinite encodings)

Reply:
It should be relatively straightforward to implement this on top of the existing implementation. We can iterate once need for it becomes more evident.